homeassistant/components/epic_games_store/calendar.py

```diff
@@ -0,0 +1,108 @@
+"""Calendar platform for a Epic Games Store."""
+
+from __future__ import annotations
+
+from datetime import datetime
+from typing import Any
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import EGSUpdateCoordinator
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the local calendar platform."""
+    coordinator: EGSUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+
+    entities = [
+        EGSFreeGamesCalendar(coordinator, entry.entry_id),
+        EGSDiscountGameCalendar(coordinator, entry.entry_id),
+    ]
+    async_add_entities(entities, True)
+
+
+class EGSCalendar(CalendarEntity):
+    """A calendar entity by Epic Games Store."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        coordinator: EGSUpdateCoordinator,
+        unique_id: str,
+        cal_type: str,
+    ) -> None:
+        """Initialize EGSCalendar."""
+        self._cal_type = cal_type
+        self._coordinator = coordinator
+        self._event: CalendarEvent | None = None
+        self._attr_name = f"Epic Games Store {cal_type.title()} Games"
+        self._attr_unique_id = f"{unique_id}-{cal_type}"
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+        return self._event
+
+    @property  # type: ignore[misc]
+    def state_attributes(self) -> dict[str, Any] | None:
+        """Return the entity state attributes."""
+        return {
+            **(super().state_attributes or {}),
+            "games": self._coordinator.data[self._cal_type],
+        }
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Get all events in a specific time frame."""
+        events: list[dict[str, Any]] = self._coordinator.data[self._cal_type]
+        return [_get_calendar_event(event) for event in events]
+
+    async def async_update(self) -> None:
+        """Update entity state with the next upcoming event."""
+        event: list[dict[str, Any]] = self._coordinator.data[self._cal_type]
+        if event:
+            self._event = _get_calendar_event(event[0])
+
+
+class EGSFreeGamesCalendar(EGSCalendar):
+    """A calendar entity by Epic Games Store."""
+
+    def __init__(
+        self,
+        coordinator: EGSUpdateCoordinator,
+        unique_id: str,
+    ) -> None:
+        """Initialize EGSCalendar."""
+        super().__init__(coordinator, unique_id, "free")
```

Comment:
If you're only going to make a constructor for cal_type, can you put it outside of the constructor?